scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/providers/TraversablesOrderedBuf.scala

```diff
@@ -0,0 +1,242 @@
+    case tpe if tpe.erasure =:= c.universe.typeOf[scala.collection.mutable.Set[Any]].erasure => TraversablesOrderedBuf(c)(buildDispatcher, tpe, DoSort, NotArray)
+
+    case tpe if tpe.erasure =:= c.universe.typeOf[Map[Any, Any]].erasure => TraversablesOrderedBuf(c)(buildDispatcher, tpe, DoSort, NotArray)
+    case tpe if tpe.erasure =:= c.universe.typeOf[scala.collection.Map[Any, Any]].erasure => TraversablesOrderedBuf(c)(buildDispatcher, tpe, DoSort, NotArray)
+    case tpe if tpe.erasure =:= c.universe.typeOf[scala.collection.mutable.Map[Any, Any]].erasure => TraversablesOrderedBuf(c)(buildDispatcher, tpe, DoSort, NotArray)
+  }
+
+  def apply(c: Context)(buildDispatcher: => PartialFunction[c.Type, TreeOrderedBuf[c.type]],
+    outerType: c.Type,
+    maybeSort: ShouldSort,
+    maybeArray: MaybeArray): TreeOrderedBuf[c.type] = {
+
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(s"fresh_$id"))
+
+    val dispatcher = buildDispatcher
+
+    val companionSymbol = outerType.typeSymbol.companionSymbol
+
+    // When dealing with a map we have 2 type args, and need to generate the tuple type
+    // it would correspond to if we .toList the Map.
+    val innerType = if (outerType.asInstanceOf[TypeRefApi].args.size == 2) {
+      val (tpe1, tpe2) = (outerType.asInstanceOf[TypeRefApi].args(0), outerType.asInstanceOf[TypeRefApi].args(1))
+      val containerType = typeOf[Tuple2[Any, Any]].asInstanceOf[TypeRef]
+      import compat._
+      TypeRef.apply(containerType.pre, containerType.sym, List(tpe1, tpe2))
+    } else {
+      outerType.asInstanceOf[TypeRefApi].args.head
+    }
+
+    val innerTypes = outerType.asInstanceOf[TypeRefApi].args
+
+    val innerBuf: TreeOrderedBuf[c.type] = dispatcher(innerType)
+
+    new TreeOrderedBuf[c.type] {
+      override val ctx: c.type = c
+      override val tpe = outerType
+      override def compareBinary(inputStreamA: ctx.TermName, inputStreamB: ctx.TermName) = {
+        val innerCompareFn = freshT("innerCompareFn")
+        val a = freshT("a")
+        val b = freshT("b")
+        q"""
+        def $innerCompareFn(a: _root_.java.io.InputStream, b: _root_.java.io.InputStream) = {
+          val $a = a
+          val $b = b
+          ${innerBuf.compareBinary(a, b)}
+        }
+        _root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.TraversableHelpers.rawCompare($inputStreamA, $inputStreamB)($innerCompareFn)
+      """
+      }
+
+      override def put(inputStream: ctx.TermName, element: ctx.TermName) = {
+        val bytes = freshT("bytes")
+        val len = freshT("len")
+        val innerElement = freshT("innerElement")
+        val cmpRes = freshT("cmpRes")
+        maybeSort match {
+          case DoSort =>
+            q"""
+          val $len = $element.size
+          $inputStream.writeSize($len)
+
+          if($len > 0) {
+            $element.toArray.sortWith { (a, b) =>
```

Comment:
let's use `Arrays.sort` with our comparator that will guarantee we are not making a copy with sortWith.